### DISCUSSION TITLE
[tvOS] Make Voiceover and Talkback fully work (including non-focusable elements)

### DISCUSSION THREAD

**Discussion by pierpo** [react-native-tvos/react-native-tvos#611] [Feature Requests]
### Description

I am working with accessibility support on my app. I tried out TalkBack, which worked as expected on `react-native-tvos`.

But VoiceOver from tvOS seems to not work properly with `react-native-tvos`.
I add `accessible` properties on my Texts, and they're still not focusable via VoiceOver. I try `accessibilityActions` with `'activate'` and I do not have my `onAccessibilityAction` callbacks triggered.

They do work properly on AndroidTV.

Any idea?

### React Native Version

0.72.6

### Output of `npx react-native info`

<details>
<summary>Output of `npx react-native info`</summary>

```
info Fetching system and libraries information...
System:
  OS: macOS 14.1
  CPU: (10) arm64 Apple M1 Pro
  Memory: 308.47 MB / 16.00 GB
  Shell:
    version: "5.9"
    path: /bin/zsh
Binaries:
  Node:
    version: 20.3.1
    path: ~/Library/Caches/fnm_multishells/24196_1699375881319/bin/node
  Yarn:
    version: 1.22.19
    path: ~/Library/Caches/fnm_multishells/24196_1699375881319/bin/yarn
  npm:
    version: 9.6.7
    path: ~/Library/Caches/fnm_multishells/24196_1699375881319/bin/npm
  Watchman:
    version: 2023.10.02.00
    path: /opt/homebrew/bin/watchman
Managers:
  CocoaPods:
    version: 1.14.2
    path: /Users/pierpo/.rbenv/shims/pod
SDKs:
  iOS SDK:
    Platforms:
      - DriverKit 23.0
      - iOS 17.0
      - macOS 14.0
      - tvOS 17.0
      - watchOS 10.0
  Android SDK:
    API Levels:
      - "28"
      - "30"
      - "31"
      - "33"
      - "34"
    Build Tools:
      - 29.0.2
      - 30.0.2
      - 30.0.3
      - 31.0.0
      - 33.0.0
      - 34.0.0
    System Images:
      - android-24 | Google APIs ARM 64 v8a
      - android-31 | Google TV ARM 64 v8a
      - android-31 | Google APIs ARM 64 v8a
    Android NDK: Not Found
IDEs:
  Android Studio: 2022.3 AI-223.8836.35.2231.10811636
  Xcode:
    version: 15.0.1/15A507
    path: /usr/bin/xcodebuild
Languages:
  Java:
    version: 17.0.6
    path: /usr/bin/javac
  Ruby:
    version: 3.0.3
    path: /Users/pierpo/.rbenv/shims/ruby
npmPackages:
  "@react-native-community/cli": Not Found
  react: Not Found
  react-native: Not Found
  react-native-macos: Not Found
npmGlobalPackages:
  "*react-native*": Not Found
Android:
  hermesEnabled: false
  newArchEnabled: false
iOS:
  hermesEnabled: false
  newArchEnabled: false
```
</details>

### Steps to reproduce

I put the following code in `packages/rn-tester/js/RNTesterAppShared.js`, and I activated VoiceOver in the tvOS settings.

```tsx
  return (
    <RNTesterThemeContext.Provider value={theme}>
      <View style={{backgroundColor: 'black', padding: 100, gap: 20}}>
        <Text
          accessible
          accessibilityLabel="read me if you can"
          style={{color: 'white', fontSize: 40}}>
          Hello, I want to be focused through accessibility but it doesn't work
          :'(
        </Text>
        <View
          accessible
          accessibilityLabel="try to focus this box if you can"
          style={{backgroundColor: 'green', height: 50, width: 50}}
        />
        <TouchableOpacity
          accessible
          accessibilityRole="button"
          accessibilityActions={[{name: 'activate'}]}
          onAccessibilityAction={() => console.log('did I work? do I didnt')}
          style={{backgroundColor: 'grey', padding: 10}}>
          <Text style={{color: 'white', fontSize: 40}}>FIRST BUTTON</Text>
        </TouchableOpacity>
        <TouchableOpacity
          accessible
          accessibilityRole="button"
          accessibilityActions={[{name: 'activate'}]}
          onAccessibilityAction={() => console.log('did I work? do I didnt')}
          style={{backgroundColor: 'grey', padding: 10}}>
          <Text style={{color: 'white', fontSize: 40}}>SECOND BUTTON</Text>
        </TouchableOpacity>
      </View>
    </RNTesterThemeContext.Provider>
  );
```

### Snack, screenshot, or link to a repository

In this example, I can only focus the two TouchableOpacity, and that's it. I would expect the green square and the text to be focusable as well.

<img width="1502" alt="image" src="https://github.com/react-native-tvos/react-native-tvos/assets/6313316/3e1757fe-fa16-4b39-90c4-3de4382854d9">

**Comment by douglowder** [comment]
I'll take a look.

**Comment by douglowder** [comment]
In general, `Text` and `View` are not focusable on Apple TV.

I'll look at making this work, but this is definitely at the level of a feature add, not a quick bug fix, unfortunately.

It seems that I need to make sure that it works with both "Navigation" and "Exploration" modes as described in https://support.apple.com/guide/tv/use-voiceover-atvbfa4ff6cd/tvos

**Comment by GuillaumeEgret** [comment]
I'm working on a skill to help developers making their apps accessible. To do so I need the voice over to be working. I opened a PR to fix it https://github.com/react-native-tvos/react-native-tvos/pull/1120 :)

**Comment by cgoldsby** [comment]
I think this feature request diverges too much from native tvOS behavior, so I have trouble justifying its inclusion in react-native-tvos.

`View` and `Text` are not focusable by default on tvOS. Focus behavior should not be tied to accessibility. VoiceOver does not make things focusable natively; when VoiceOver is enabled, focus behavior should not change.

I tested this in a native tvOS app. A `UILabel` with `isAccessibilityElement = true` appears in the VoiceOver Item Chooser, but VoiceOver swipe navigation still cannot reach it. Only views where `canBecomeFocused` returns `true` (and `isUserInteractionEnabled = true`) are reachable by swipe. This is how tvOS works; react-native-tvos mirrors that correctly.

My core issue with the PR: `canBecomeFocused` should not be conditioned on `UIAccessibilityIsVoiceOverRunning()`. That couples focus behavior to assistive technology state, which is not how native tvOS apps work.

One workaround: toggle `focusable` when VoiceOver is active. `AccessibilityInfo` already provides everything you need.

```tsx
const [voiceOverOn, setVoiceOverOn] = useState(false);

useEffect(() => {
  AccessibilityInfo.isScreenReaderEnabled().then(setVoiceOverOn);
  const sub = AccessibilityInfo.addEventListener(
    'screenReaderChanged',
    setVoiceOverOn,
  );
  return () => sub.remove();
}, []);

// View becomes focusable only when VoiceOver is running
<View
  accessible
  focusable={voiceOverOn}
  accessibilityLabel="Read me">
  <Text>Hello, I am a plain text</Text>
</View>
```

Note: `focusable` works on `<View>` but has no effect on `<Text>` (because `RCTParagraphComponentView` does not check `isTVSelectable`). Wrap text in a `<View>` or `<Pressable>`. I am open to adding `focusable`/`isTVSelectable` support to `RCTParagraphComponentView` if that would help.
